File: 872. Leaf-Similar Trees/Leaf-Similar Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        ArrayList<TreeNode> one = new ArrayList<TreeNode>();
        ArrayList<TreeNode> two = new ArrayList<TreeNode>();
        one.add(root1);
        two.add(root2);
        ArrayList<Integer> leafOne = new ArrayList<Integer>();
        ArrayList<Integer> leafTwo = new ArrayList<Integer>();
        TreeNode current;
        while (one.size() > 0){
            current = one.remove(one.size()-1);
            if (current.left == null && current.right == null){
                leafOne.add(current.val);
            }
            else{
                if (current.left != null){
                    one.add(current.left);
                }
                if (current.right != null){
                    one.add(current.right);
                }
            }
        }
        while (two.size() > 0){
            current = two.remove(two.size()-1);
            if (current.left == null && current.right == null){
                leafTwo.add(current.val);
            }
            else{
                if (current.left != null){
                    two.add(current.left);
                }
                if (current.right != null){
                    two.add(current.right);
                }
            }
        }
        if(leafOne.size() != leafTwo.size()){
            return false;
        }
        for (int i = 0; i < leafOne.size(); i++){
            if (leafOne.get(i) != leafTwo.get(i)){
                return false;
            }
        }
        return true;
    }
}

File: 872. Leaf-Similar Trees/Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<TreeNode*> one;
        vector<TreeNode*> two;
        vector<int> leafOne;
        vector<int> leafTwo;
        one.push_back(root1);
        two.push_back(root2);
        TreeNode current;
        while(!one.empty()){
            current = *one[one.size()-1];
            one.pop_back();
            if (current.left == NULL && current.right == NULL){
                leafOne.push_back(current.val);
            }
            else{
                if (current.left != NULL){
                    one.push_back(current.left);
                }
                if (current.right != NULL){
                    one.push_back(current.right);
                }
            }
        }
        while(!two.empty()){
            current = *two[two.size()-1];
            two.pop_back();
            if (current.left == NULL && current.right == NULL){
                leafTwo.push_back(current.val);
            }
            else{
                if (current.left != NULL){
                    two.push_back(current.left);
                }
                if (current.right != NULL){
                    two.push_back(current.right);
                }
            }
        }
        if (leafOne.size() != leafTwo.size()){
            return false;
        }
        for(int i = 0; i < leafOne.size(); i++){
            if (leafOne[i] != leafTwo[i]){
                return false;
            }
        }
        return true;
    }
};
